applications/nrf_desktop/src/hw_interface/selector_hw.c

```diff
@@ -0,0 +1,354 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <gpio.h>
+
+#include "selector_hw_def.h"
+
+#include "event_manager.h"
+#include "selector_event.h"
+#include "power_event.h"
+
+#define MODULE selector_hw
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_SELECTOR_HW_LOG_LEVEL);
+
+#define DEFAULT_STATE ARRAY_SIZE(pins)
+
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+};
+
+struct selector {
+	struct gpio_callback gpio_cb[ARRAY_SIZE(port_map)];
+	struct k_delayed_work work;
+	u8_t state;
+	u8_t id;
+	u8_t pin_cnt;
+};
+
+static struct device *gpio_dev[ARRAY_SIZE(port_map)];
+static struct selector selectors[ARRAY_SIZE(pins)];
+static enum state state;
+
+
+static void selector_event_send(const struct selector *selector)
+{
+	BUILD_ASSERT_MSG(ARRAY_SIZE(pins) <= UCHAR_MAX,
+			 "Maximum number of selectors is 255");
+	__ASSERT_NO_MSG(selector->state < ARRAY_SIZE(pins[selector->id]));
+
+	struct selector_event *event = new_selector_event();
+
+	event->selector_type = SELECTOR_TYPE_HW;
+	event->selector_id = selector->id;
+	event->state = selector->state;
+
+	EVENT_SUBMIT(event);
+}
+
+static int read_state(struct selector *selector)
+{
+	u32_t value;
+	const struct pin *sel_pins = pins[selector->id];
+
+	for (size_t i = 0; i < selector->pin_cnt; i++) {
+		int err = gpio_pin_read(gpio_dev[sel_pins[i].port],
+					sel_pins[i].pin,
+					&value);
+
+		if (err) {
+			LOG_ERR("Cannot read value from port:%" PRIu8
+				" pin: %" PRIu8, sel_pins[i].port,
+				sel_pins[i].pin);
+			return err;
+		}
+
+		if (value) {
```

Comment:
I do not like such code when you return from the function  in the middle of a loop. Change that to break the loop and if value is provided then set selector state and call selector_event_send once after that.

Reply:
ok

Reply:
done